cpp/tests/sampling/random_walks_profiling.cu

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuda_profiler_api.h>
+#include <gtest/gtest.h>
+
+#include <utilities/base_fixture.hpp>
+#include <utilities/high_res_timer.hpp>
+#include <utilities/test_utilities.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/random.h>
+
+#include <algorithms.hpp>
+#include <experimental/random_walks.cuh>
+#include <graph.hpp>
+
+#include <raft/handle.hpp>
+#include <raft/random/rng.cuh>
+
+#include <algorithm>
+#include <iterator>
+#include <limits>
+#include <numeric>
+#include <vector>
+
+namespace {  // anonym.
+template <typename vertex_t, typename index_t>
+void fill_start(raft::handle_t const& handle,
+                rmm::device_uvector<vertex_t>& d_start,
+                index_t num_vertices)
+{
+  index_t num_paths = d_start.size();
+
+  thrust::transform(rmm::exec_policy(handle.get_stream())->on(handle.get_stream()),
+                    thrust::make_counting_iterator<index_t>(0),
+                    thrust::make_counting_iterator<index_t>(num_paths),
+
+                    d_start.begin(),
+                    [num_vertices] __device__(auto indx) { return indx % num_vertices; });
+}
+}  // namespace
+
+struct RandomWalks_Usecase {
+  std::string graph_file_full_path{};
+  bool test_weighted{false};
+
+  RandomWalks_Usecase(std::string const& graph_file_path, bool test_weighted)
+    : test_weighted(test_weighted)
+  {
+    if ((graph_file_path.length() > 0) && (graph_file_path[0] != '/')) {
+      graph_file_full_path = cugraph::test::get_rapids_dataset_root_dir() + "/" + graph_file_path;
+    } else {
+      graph_file_full_path = graph_file_path;
+    }
+  };
+};
+
+class Tests_RandomWalksProfiling : public ::testing::TestWithParam<RandomWalks_Usecase> {
+ public:
+  Tests_RandomWalksProfiling() {}
+  static void SetupTestCase() {}
+  static void TearDownTestCase() {}
+
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  template <typename vertex_t, typename edge_t, typename weight_t>
+  void run_current_test(RandomWalks_Usecase const& configuration)
```

Comment:
I don't think we should be using gtest this way.  I'd rather see a simple executable that does this kind of benchmarking, or add to the existing python benchmarks. The latter adds more value since they'll show up in reports and can help catch regressions.

Reply:
I don't understand what is the course of action, here. Can you point to an example of how should this be tested? Do we have what you want in our code already? If so, please point me there. If not, please provide a code snippet example.

This is done in many places in the code, including recent `egonet`. I'm not sure why it was okay there but not here. At which point I'd like to point out this is not a PR for revamping the way we test. That is a major code re-org beyond the scope not only of this PR, but also of this release, as I'm told.

If a simple fix is possible for this particular profiling, I'd like to see an example, please. Thank you.

Reply:
Sorry if I wasn't clear before. The issue here is that you're writing a benchmark but using gtest, which is not what it's intended for.  You're not actually testing anything (ie. there's no assertions).  I'm also concerned that this will propagate an anti-pattern (ie. writing gtests that aren't actually tests), and we'll have a large refactoring ahead of us. (BTW, the egonet tests are actual tests that have assertions).

For benchmarking, I'd consider either doing it in python (as done [here](https://github.com/rapidsai/cugraph/tree/branch-0.20/benchmarks)), or using gbench, or even a separate executable for C++. For profiling, we should also look into using [nvtx](https://docs.nvidia.com/gameworks/content/gameworkslibrary/nvtx/nvidia_tools_extension_library_nvtx.htm) which is what cudf and cuml are starting to use so we're consistent with RAPIDS.  Some of these suggestions may be considered out-of-scope for this PR though, so my recommendation would be to simply make this a (non-gtest) standalone executable.  Since you're not using many gtest features it shouldn't be a big change, and I can work with you offline if that doesn't make sense.  Another option would be to simply add the CUDA profiling and timer calls to your other gtests, so that way running a test will actually test something and it will be consistent with what egonet, etc. currently do.

I'm looking forward to more benchmarks and also profiling, especially ones that exercise our C++ library directly, and I think we can use this as an opportunity to start that practice.

Reply:
It was specifically requested to add C++ profiling. I can certainly add testing predicates to qualify this as a test. 

Also, I don't mind introducing `gbench` in this PR, provided is not to time consuming and it doesn't need tons of boilerplate coding. This is meant as a _quick_ enhancement PR (emphasis on "quick") as we have more pressing priorities than `gbench`, right now. But, I will probably need some help with `gbench` as I have no prior experience with it. If it's something simple I can add that to this PR. If not, this would probably go into our re-org goals in following release(s).

Reply:
I am considering making it a separate executable. Do we have an example in our code you could point me to? If not what is the `cmake` machinery involved? Which `cmake` must be changed and how?

Reply:
I think that's the best approach for this as well.  I think that'll give us more flexibility too since we won't be constrained by gtest's testing assumptions.  I can help with this but I unfortunately don't have an example to point you to.  I'll chat with you offline and then you or we can update the PR (I don't think it'll be a big change).